java/feathub-connectors/flink-connectors/flink-connector-prometheus/src/main/java/com/alibaba/feathub/flink/connectors/PrometheusSinkFunction.java

```diff
@@ -0,0 +1,366 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.feathub.flink.connectors;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
+import org.apache.flink.table.catalog.Column;
+import org.apache.flink.table.catalog.ResolvedSchema;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.DecimalData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeFamily;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.MapType;
+import org.apache.flink.util.StringUtils;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.type.TypeReference;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.SimpleCollector;
+import io.prometheus.client.exporter.PushGateway;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.DELETE_ON_SHUTDOWN;
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.GROUPING_KEY;
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.HOST_URL;
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.JOB_NAME;
+
+/**
+ * A {@link org.apache.flink.streaming.api.functions.sink.SinkFunction} that writes to a Prometheus
+ * PushGateway. Note that the SinkFunction only support writing numeric columns or map columns that
+ * map from string to numeric values. Each column can be configured with a json map via the column
+ * comment. Below are the configuration of the column:
+ *
+ * <ul>
+ *   <li>metric_name: This specifies the metric name under which the column value is reported. If
```

Comment:
I am wondering which use-case requires this capability of renaming the metric. Can you describe such a use-case?

It is mentioned in the PushGateway README that "Metrics of the same name must have the same type". So we probably have to encode some of the "tags" in the metric name so that we can derive metrics with different type from the same feature anyway.

Reply:
This is required by the Feathub as Feathub metrics all have the same name, such as count, ratio, etc. They are differentiated only by the label value. And Flink table cannot have two columns with the same name. Therefore, we need to have a way to map different columns to the same metric name.